### ISSUE TITLE
Having multiple pods with their vendored frameworks in the same parent folder results in malformed archive due to bcsymbolmap files

### ISSUE THREAD

**Issue by acecilia** [CocoaPods/CocoaPods#10385]
* [x] I've read and understood the [*CONTRIBUTING* guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report

## What did you do?
Make two podspec files with one vendored framework each. Both vendored frameworks are inside the same parent folder, and next to them are the corresponding `dsym` and `bcsymbolmap` files (binaries are created using carthage). Then I add them as dependencies of an app, and tried to archive and export it

## What did you expect to happen?
The archiving and exporting succeeds

## What happened instead?

Archive succeeded, but the generated archive is malformed and cant be exported:
```
...
** ARCHIVE SUCCEEDED ** [4.986 sec]

xcodebuild -exportArchive -archivePath App.xcarchive -exportPath App.ipa -exportOptionsPlist exportOptions.plist
error: archive at path '/Users/.../cocoapods_bcsymbolmap2/App.xcarchive' is malformed
** EXPORT FAILED **
```

## CocoaPods Environment

Cocoapods 1.10.0
Catalina 10.15.7

## Project that demonstrates the issue

1. Clone https://github.com/acecilia/cocoapods_bcsymbolmap2
2. Set a valid DEVELOPMENT_TEAM  in https://github.com/acecilia/cocoapods_bcsymbolmap2/blob/main/project.yml#L11
3. Run `make export_app` from the root

## The issue explained:

Seems the issue is caused by the `bcsymbolmap` files being considered multiple times in the build graph: each of the pods is including all the bcsymbolmap files around (not just its own!). For example, the following steps generate a valid archive:
1. Remove the `bcsymbolmap` files
2. Remove the `Pods` folder
3. Run `make export_app` again

## Workaround

Add a post_install step to modify the files used during the copy dsym build phase, so the `bcsymbolmap` files are only added once:
```ruby
def adjust_copy_dsyms_phase(installer)
    [
        installer.development_pod_targets.map { |t| t.copy_dsyms_script_input_files_path },
        installer.development_pod_targets.map { |t| t.copy_dsyms_script_output_files_path },
        installer.development_pod_targets.map { |t| t.copy_dsyms_script_path }
    ].each { |files|
        already_included_bcsymbolmap_files = []
        files.each { |file|
            next if not File.exist?(file) 
            already_included_bcsymbolmap_files += remove_duplicated_bcsymbolmap_lines(file, already_included_bcsymbolmap_files)
        }
    }
end

def remove_duplicated_bcsymbolmap_lines(path, already_included_bcsymbolmap_files)
    lines = []
    bcsymbolmap_ids = []

    for line in File.readlines(path).map { |line| line.strip }
        if line.include? ".bcsymbolmap"
            # Workaround for https://github.com/CocoaPods/CocoaPods/issues/10373 
            next if lines.include? line

            # Workaround for https://github.com/CocoaPods/CocoaPods/issues/10385
            bcsymbolmap_id = File.basename(line.tr('"', ''), ".bcsymbolmap")
            if not already_included_bcsymbolmap_files.include? bcsymbolmap_id
                lines.append(line)
                bcsymbolmap_ids.append(bcsymbolmap_id)
            end
        else
            lines.append(line)
        end
    end
  
    File.open(path, "w+") do |f|
        f.puts(lines)
    end

    return bcsymbolmap_ids
end

post_install do |installer|
    adjust_copy_dsyms_phase(installer)
end

```

**Comment by dnkoutso** [comment]
```Touch /Users/dimitris/Library/Developer/Xcode/DerivedData/TestProject-aqmhhvkayugdjfgwjwvcmwibmlwr/Build/Intermediates.noindex/ArchiveIntermediates/App/InstallationBuildProductsLocation/Applications/App.app (in target 'App' from project 'TestProject')
    cd /Users/dimitris/Development/ios/scrap/cocoapods_bcsymbolmap2
    /usr/bin/touch -c /Users/dimitris/Library/Developer/Xcode/DerivedData/TestProject-aqmhhvkayugdjfgwjwvcmwibmlwr/Build/Intermediates.noindex/ArchiveIntermediates/App/InstallationBuildProductsLocation/Applications/App.app

** ARCHIVE SUCCEEDED ** [3.504 sec]```

I am able to. build and archive the project even with cocoapods 1.10 and 1.11.

**Comment by acecilia** [comment]
@dnkoutso I tried the sample project with cocoapods `1.11.0.beta.2` and it keeps failing:
```
** ARCHIVE SUCCEEDED ** [3.009 sec]

xcodebuild -exportArchive -archivePath App.xcarchive -exportPath App.ipa -exportOptionsPlist exportOptions.plist
error: archive at path '/Users/.../cocoapods_bcsymbolmap2/App.xcarchive' is malformed
** EXPORT FAILED **

make: *** [export_app] Error 65
```

Note that archive succeeds (see the log of the first message of the issue): **what fails is exporting**.

My machine:
```
xcodebuild -version
Xcode 12.3
Build version 12C33


MacOS big sur 11.4 (20F71)
```

**Comment by acecilia** [comment]
@dnkoutso I think this is still an issue, could you reopen 🙏

**Comment by dnkoutso** [comment]
did you try with final 1.11.0?

**Comment by acecilia** [comment]
I just tried version `1.11.0`, archive succeeds but export keeps failing. I updated the sample with cocoapods `1.11.0` and I removed the need for signing. Could you please try and see if you can reproduce the **export failure** on your side? 
Just run: `git clone git@github.com:acecilia/cocoapods_bcsymbolmap2.git && cd cocoapods_bcsymbolmap2 && make export_app`

**Comment by dnkoutso** [comment]
yeah, I wonder why this happens...will try this week hopefully.

**Comment by dnkoutso** [comment]
interesting, not sure why this would happen to be honest during exporting. Is there more logging to output the error message?

I'd expect we see a duplicate output path message but we dont so I again not sure whats the underlying issue?

**Comment by dnkoutso** [comment]
are you intentionally sharing those bcsymbolmaps across these two pods? I wonder if we do that accidentally by globing too much?

Do you expect both of these pods to use the same set of bcsymbolmaps?

**Comment by dnkoutso** [comment]
oooh @acecilia can you output the carthage build into a sub folder?

like 
```
Carthage/Build/iOs/DiffableDataSources/DiffableDataSources.framework
Carthage/Build/iOs/DifferenceKit/DifferenceKit.framework
```

and place the bcsymbol maps to each corresponding folder? CocoaPods does the following;

```ruby
dirname = framework_path.dirname
bcsymbolmap_paths = if dirname.exist?
                      Dir.chdir(dirname) do
                        Dir.glob('*.bcsymbolmap').map do |bcsymbolmap_file_name|
                          bcsymbolmap_path = dirname + bcsymbolmap_file_name
                          "${PODS_ROOT}/#{bcsymbolmap_path.relative_path_from(sandbox.root)}"
                        end
                      end
                    end
```

So in this case outputing the frameworks in the same folder would catch the bcsymbolmaps for both framework targets.

**Comment by acecilia** [comment]
> Is there more logging to output the error message?

Nope

> I'd expect we see a duplicate output path message but we dont so I again not sure whats the underlying issue?

Cant tell 😢 

> are you intentionally sharing those bcsymbolmaps across these two pods?

Nope

> Do you expect both of these pods to use the same set of bcsymbolmaps?

Nope. Each framework has its own bcsymbolmap files, its just that they are under the same parent folder. I am not sure if cocoapods can distinguish somehow which bcsymbolmap files are associated with which framework

> So in this case outputing the frameworks in the same folder would catch the bcsymbolmaps for both framework targets.

Yes, seems that this is the issue. Ideally each framework should only catch its own bcsymbolmap files

> can you output the carthage build into a sub folder? and place the bcsymbol maps to each corresponding folder?

I would love to do that, but no, Carthage does not support that

**Comment by dnkoutso** [comment]
yes unfortunately right now cocoapods wont be able to distinguish that. There is no DSL for bcsymbol maps. You'd have to somehow specify the output to a subfolder or move the files your self after the carthage build runs if you know which map is for which framework.

**Comment by acecilia** [comment]
Got it. Then I will keep using the workaround I pasted in the issue description 🙌  Thanks! 🙏

**Comment by dnkoutso** [comment]
yes it might be easier and cleaner to move things around to avoid all this editing though!

@acecilia should I close this? I do not think we will add a DSL anytime soon... the workaround for folks should be sufficient.

**Comment by acecilia** [comment]
> how do you know??

Yeah, not sure there is a way to know. And a DSL would not 100% solve this issue anyway, it would just pass the responsibility to the author of the podspec.

Since the workaround above is not something cocoapods can implement, only way I can see to solve this would be match frameworks with bcsymbolmap files from inside cocoapods. But not sure if that is even possible.

Other alternative would be that Carthage allows to place framework in subfolders. Very unlikely I would say

**Comment by acecilia** [comment]
Thanks for the follow up! ⭐
